torch/csrc/jit/interpreter.cpp

```diff
@@ -1,42 +1,208 @@
+    }
+    autograd::variable_list v_outputs = func->apply(v_inputs);
+    for(auto & output : v_outputs) {
+      outputs.push_back(builder.addOutput(output));
+    }
+    builder.writeTo(outputs);
+  };
+}
+
+Operation createEvalCallback(CppOp * op) {
+  bool has_handle_output = hasHandleOutput(op);
+  return [=](const refcounted_list & inputs,
+             refcounted_list & outputs) {
+    Handle * handle_in = dynamic_cast<Handle*>(inputs.back());
+    JIT_ASSERT(handle_in);
+    HandleBuilder builder(has_handle_output);
+    auto& engine = autograd::python::PythonEngine::getDefaultEngine();
+    autograd::variable_list v_inputs;
+    for(size_t i = 0; i < inputs.size() - 1; i++) {
+      v_inputs.push_back(builder.addInput(inputs[i], op->var_flags[i]));
+    }
+    autograd::Engine::pre_callback_map callbacks;
+    callbacks.emplace(handle_in->forward_inputs.get(), [&](autograd::Function * _unused, autograd::variable_list & values) -> bool {
+      for(auto & v : values) {
+        outputs.push_back(builder.addOutput(v));
+      }
+      return false; // stop output and do not run DummyFunction
+    });
+    // node handle_in->use_count() == 1 means that we are guarenteed that we have the only
+    // only copy of the backward pass, but it is not clear that it is safe even in that case
+    // if use_count() > 1 then it is possible it drops to 1 during this execute,
+    // however, that is safe if conservative since it will get freed when the handles die.
+    engine.execute(handle_in->forward_outputs, v_inputs, handle_in->use_count() > 1, callbacks);
+    builder.writeTo(outputs);
+  };
+}
+
 using tensor_list = std::vector<at::Tensor>;
-using Callback = std::function<void(const tensor_list &, tensor_list &)>;
 // Returns a function implementing functionality of a given node,
 // or nullptr if it's a no-op for autograd.
-Callback getCallback(Node *node) {
+Operation getOperation(jit::Node *node) {
   IR_IFM(node, PythonOp)
-    throw NotImplementedException();
+    return createPythonCallback(value);
   IR_ELSEIFM(CppOp)
-    throw NotImplementedException();
-  IR_ELSEIF(Select)
-    barf("getCallback() on select?");
+    if(dynamic_cast<autograd::Eval*>(value->fn.get())) {
+      return createEvalCallback(value);
+    } else {
+      return createCppCallback(value);
+    }
   IR_ELSEIF(FusionGroup)
 #ifdef WITH_CUDA
     auto fusion_fn = sharedFusionCompiler().getOrCompile(*value->g(kSubgraph));
-    return [fusion_fn](const tensor_list & inputs, tensor_list & outputs) {
+    return [fusion_fn](const refcounted_list & inputs, refcounted_list & outputs) {
       autograd::profiler::RecordFunction record("FusionGroup");
-      fusion_fn->launch(inputs, outputs);
+      tensor_list tinputs, toutputs;
+      tinputs.reserve(inputs.size());
+      for(auto & i : inputs) {
+        tinputs.push_back(unsafeTensorBorrow(i));
+      }
+      fusion_fn->launch(tinputs, toutputs);
```

Comment:
This could accept those temporary lists right?

Reply:
Not easily -- You _cannot_ pass a TensorTemporary to a thing that takes std::vector<Tensor> because it will slice off the Temporary and then (incorrectly) delete the tensor. I am going to change around the implementation of TensorTemporary to make its use safer.